### ISSUE TITLE
Build etherPad component

### ISSUE THREAD

**Issue by JamesPHoughton** [Watts-Lab/deliberation-empirica#521]
Create a new component in the `client/src/components` directory that will embed an etherpad instance in our experiment. Then add the appropriate connections in the generic `Element.jsx` class. We should call this a 'sharedNotepad` component. 

Steps:
1. get in touch with mark and his RA and learn how the etherpad component works, and write your notes in this issue so that I can know too
2. Develop the component
3. Write cypress tests in omnibus test
  - modify the treatment file for the omnibus test to include the etherpad component
  - modify the tests themselves:
    - player 1 makes a change
    - player 2 sees the change that player 1 made
    - player 2 makes a change
    - player 1 sees the change player 2 made

**Comment by gumillie222** [comment]
Here is the Etherpad [document](https://etherpad.org/doc/v1.8.4/)

A helpful image in understanding the structure:
<img width="1459" alt="Screenshot 2023-07-14 at 1 54 39 AM" src="https://github.com/Watts-Lab/deliberation-empirica/assets/121700352/084c4770-2829-4cd7-bb62-eb60f7470b36">

**Comment by JamesPHoughton** [comment]
A few of our current RAs have used it and we have plenty of example tasks in Empirica 1 that use it (e.g. https://github.com/Watts-Lab/multi-task-empirica/blob/main/multi-task-app/customTasks/MoralReasoning/game/index.jsx) with a component we developed (https://github.com/Watts-Lab/multi-task-empirica/blob/main/multi-task-app/customTasks/Etherpad.jsx).

**Comment by kepstein23** [comment]
@gumillie222 and I decided on `groupID + stageName` as a naming scheme for the padID. This makes it so that each group can only see one pad per stage.  This can be changed by making a more complicated naming scheme later on which would be done by changing how padID is set in Elements.jsx.

**Comment by gumillie222** [comment]
We currently have shareNotepad working as part of Qualtrics.jsx (it was running into issues when we separated it out). Both players can edit the pad and see notes from each other.

TO-DO:
1. clear interval on stage submit
2. save pad on stage submit
3. better way for padID
4. separate from Qualtrics.jsx

**Comment by JamesPHoughton** [comment]
Is there a way to get the final state of the pad, and save it to the player object, like we do with prompts? We'll probably also need to make changes to our data export to make sure we save the data. It would be nice to save both the final text and also either a history of edits, or have the final text annotated somehow to show who wrote each part. How would we do that?

**Comment by kepstein23** [comment]
> Is there a way to get the final state of the pad, and save it to the player object, like we do with prompts? 

Yes, I believe theres a function in the multi-task code that gets the text from the etherpad api.  We were having issues getting node-fetch to work, but I think we'll need to do this in onStageSubmit (and save it to the player object).

>It would be nice to save both the final text and also either a history of edits, or have the final text annotated somehow to show who wrote each part. How would we do that?

I'm not sure how to do this - we could save them all to the stage object with a timestamp.  I think multi-task was keeping track of who was editing with text color so we could do this as well. If we didn't want the display to change we could maybe tag the text with the player id, then save each player's text individually? There might be an easier way to do this as well.

**Comment by gumillie222** [comment]
updated To-dos:
1. write cypress
2. figure out a way to distinguish who wrote what in the saved data

**Comment by JamesPHoughton** [comment]
An etherpad is essentially a shared open response. If we add a "shared" option to the prompt element, then we can put in an etherpad instead of a regular text box. Then we can use the prompt options to specify the etherpad default text.

**Comment by JamesPHoughton** [comment]
Problem serving in safari and in cypress:
<img width="819" alt="image" src="https://github.com/Watts-Lab/deliberation-empirica/assets/4304478/c9691143-4436-44b2-a0ec-0c584b135641">
Has to do with same-origin policy for setting cookies: Safari won't let embedded pages set cookies unless you have previously visited the other domain and set the cookie.

- https://gist.github.com/iansltx/18caf551baaa60b79206
- https://www.codeproject.com/Articles/5330276/Cross-Domain-Embedding-Making-Third-Party-Cookies

@markwhiting - did you deal with this in your other implementation of etherpad? or do you just require chrome?

**Comment by JamesPHoughton** [comment]
I guess I could have our load balancer forward traffic from `study.deliberation-lab.org/etherpad` to the correct host, but that doesn't help with development...

**Comment by JamesPHoughton** [comment]
The other strategy is to set up a page on the etherpad server that we can send users to and have them bounce back from?

**Comment by markwhiting** [comment]
We haven't resolved this and for now just strongly recommend Chrome, however, it's not actually an issue with Safari specifically, it's an issue with security preferences, of which Safari has the strongest defaults. I've found that users of other browsers including Chrome run into this if they have tweaked their security settings to be stronger than default. 

I have not tried to solve it, but our students and IT didn't have a good work around last we checked.
